sgv2-graphqlapi/src/main/java/io/stargate/sgv2/graphql/web/resources/StargateGraphqlContext.java

```diff
@@ -0,0 +1,153 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.stargate.sgv2.graphql.web.resources;
+
+import com.google.common.base.MoreObjects;
+import io.stargate.proto.QueryOuterClass.BatchParameters;
+import io.stargate.proto.QueryOuterClass.BatchQuery;
+import io.stargate.proto.QueryOuterClass.Query;
+import io.stargate.proto.QueryOuterClass.QueryParameters;
+import io.stargate.proto.QueryOuterClass.Response;
+import io.stargate.sgv2.common.grpc.StargateBridgeClient;
+import io.stargate.sgv2.graphql.schema.CassandraFetcher;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.atomic.AtomicReference;
+import javax.servlet.http.HttpServletRequest;
+
+public class StargateGraphqlContext {
+
+  private final HttpServletRequest httpRequest;
+  private final StargateBridgeClient bridge;
+  private final GraphqlCache graphqlCache;
+  private final BatchContext batchContext = new BatchContext();
+
+  private volatile boolean overloaded;
+
+  public StargateGraphqlContext(
+      HttpServletRequest httpRequest, StargateBridgeClient bridge, GraphqlCache graphqlCache) {
+    this.httpRequest = httpRequest;
+    this.bridge = bridge;
+    this.graphqlCache = graphqlCache;
+  }
+
+  public StargateBridgeClient getBridge() {
+    return bridge;
+  }
+
+  public BatchContext getBatchContext() {
+    return batchContext;
+  }
+
+  /**
+   * Records the fact that at least one CQL query in the current execution failed with an OVERLOADED
+   * error. This will be translated into an HTTP 429 error at the resource layer.
+   */
+  public void setOverloaded() {
+    this.overloaded = true;
+  }
+
+  public boolean isOverloaded() {
+    return overloaded;
+  }
+
+  /**
+   * Encapsulates logic to add multiple queries contained in the same operation that need to be
+   * executed in a batch.
+   */
+  public static class BatchContext {
+    private final List<BatchQuery> queries = new ArrayList<>();
+    private int operationCount;
+    private final CompletableFuture<Response> executionFuture = new CompletableFuture<>();
+    private final AtomicReference<QueryParameters> parameters = new AtomicReference<>();
+
+    public CompletableFuture<Response> getExecutionFuture() {
+      return executionFuture;
+    }
+
+    public synchronized List<BatchQuery> getQueries() {
+      return queries;
+    }
+
+    public void setExecutionResult(CompletionStage<Response> response) {
+      response
+          .thenApply(executionFuture::complete)
+          .exceptionally(executionFuture::completeExceptionally);
+    }
+
+    public void setExecutionResult(Exception ex) {
+      executionFuture.completeExceptionally(ex);
+    }
+
+    public synchronized int add(Query query) {
```

Comment:
Because it mutates non-atomic state and can be called concurrently by multiple fetchers (in the `@atomic` example I posted in one of my other answers, the fetchers run concurrently).

Reply:
OK, maybe we should move to concurrent data type and not have the synchronized here, but anyway